src/models/Review.js

```diff
@@ -1,16 +1,38 @@
 const mongoose = require('mongoose');
+const Product = require('./Product');
 
-const reviewSchema = new mongoose.Schema({
-    rating: {
-        type: Number,
-        min: 0,
-        max: 5,
-        default: 1,
+const reviewSchema = new mongoose.Schema(
+    {
+        rating: {
+            type: Number,
+            min: 0,
+            max: 5,
+            default: 1,
+        },
+        comment: {
+            type: String,
+        },
     },
-    comment: {
-        type: String,
+    { timestamps: true, versionKey: false },
+);
+
+// pre-hook on remove : when review is deleted then it should be removed from product
+// and it's contribution to product's average rating should also be removed
+reviewSchema.pre(
+    'deleteOne',
+    { document: true, query: false },
+    async function (next) {
+        const product = await Product.findOne({ reviews: this._id });
+        // console.log(`product : \n ${product}`);
+        product.rating = (product.rating * product.reviews.length - this.rating)
+      / (product.reviews.length - 1);
+        product.reviews = product.reviews.filter(
+            (review) => !review._id.equals(this._id),
+        );
```

Comment:
Use mongo's `$pull` to delete any value from array.

Reply:
I think then I have to use the update() on Product model rather than save()

Reply:
Yes lets use update then, code will be more clean that way, also save uses the update under the hood for existing document

Reply:
I have a query, when a review gets deleted, then we have to do two things

1. remove the reference of review in products.reviews array => which we can do at db level something like this
`
await Product.updateOne(
  {reviews: this._id},
  {$pull: {reviews: this._id}}
)
`
2. re-calculate the average rating of the product for which we would have to fetch the doc into the app -> calculate avg rating -> save it back to mongo
`
product.setAverageRating();
product.save()
`
So my query is we have to perform two changes...one can be done using update query at db level...and other needs to be done by explicitly fetching it in app and then saving it

Can we do both at the db level...as I'm already fetching the doc into the application, I'm filtering out the review at the application level itself.